x-pack/plugins/fleet/server/routes/security.ts

```diff
@@ -165,22 +93,199 @@ export async function getAuthzFromRequest(req: KibanaRequest): Promise<FleetAuth
   });
 }
 
-export type RouterWrapper = <T extends RequestHandlerContext>(route: IRouter<T>) => IRouter<T>;
+function hasRequiredFleetAuthzPrivilege(
+  authz: FleetAuthz,
+  {
+    fleetAuthz,
+    fleetAllowFleetSetupPrivilege,
+  }: { fleetAuthz?: FleetAuthzRouteConfig; fleetAllowFleetSetupPrivilege?: boolean }
+): boolean {
+  if (!checkSecurityEnabled()) {
+    return false;
+  }
 
-interface RouterWrappersSetup {
-  require: {
-    superuser: RouterWrapper;
-    fleetSetupPrivilege: RouterWrapper;
-  };
+  if (fleetAllowFleetSetupPrivilege) {
+    // Temporary while agent call the setup API
+    if (authz.fleet.setup) {
+      return true;
+    }
+  }
+
+  const missingAuthz = [];
+
+  if (fleetAuthz && typeof fleetAuthz.fleet !== 'undefined') {
+    for (const key of fleetAuthz.fleet) {
+      if (!authz.fleet[key as keyof FleetAuthz['fleet']]) {
+        missingAuthz.push(`fleet.${key}`);
+      }
+    }
+  }
+
+  if (fleetAuthz && typeof fleetAuthz.integrations !== 'undefined') {
+    for (const key of fleetAuthz.integrations) {
+      if (!authz.integrations[key as keyof FleetAuthz['integrations']]) {
+        missingAuthz.push(`integrations.${key}`);
+      }
+    }
+  }
+
+  if (missingAuthz.length > 0) {
+    return false;
+  }
+
+  return true;
+}
+
+interface FleetAuthzRouteConfig {
+  fleet?: Array<keyof FleetAuthz['fleet']>;
+  integrations?: Array<keyof FleetAuthz['integrations']>;
```

Comment:
Would it be simpler if we made this the same shape as `FleetAuthz`? We could use a mapped type to create a type where all keys are optional and the values must be `true` (because `false` in this case doesn't make sense):

```ts
type DeepPartialTruthy<T> = {
  [P in keyof T]?: T[P] extends boolean ? true : DeepPartialTruthy<T[P]>;
};

type FleetAuthzRouteConfig = DeepPartialTruthy<FleetAuthz>;
```

Reply:
I think the array version was easy to use and to configure, is there a value to use the same shape?

Reply:
I think it's a bit cleaner and easier to understand to be able to use the same interface everywhere. That way it's obvious that these two concepts / APIs are related to one another. Another reason could be if we ever introduce nested fields this may be nice. For example `fleetAuthz.fleet.agents.canEnroll`

That said, I don't have a strong preference and this can easily be changed in the future if needed.

Reply:
You convince me I did the change :)